lib/archethic/utxo.ex

```diff
@@ -0,0 +1,146 @@
+  alias Archethic.TransactionChain.Transaction.ValidationStamp.LedgerOperations
+
+  alias Archethic.TransactionChain.Transaction.ValidationStamp.LedgerOperations.TransactionMovement
+
+  alias Archethic.TransactionChain.Transaction.ValidationStamp.LedgerOperations.UnspentOutput
+
+  alias Archethic.TransactionChain.Transaction.ValidationStamp.LedgerOperations.VersionedUnspentOutput
+
+  alias Archethic.Utils
+
+  alias Archethic.UTXO.DBLedger
+  alias Archethic.UTXO.Loader
+  alias Archethic.UTXO.MemoryLedger
+
+  require Logger
+
+  @type load_opts :: [io_transaction?: boolean()]
+
+  @spec load_transaction(Transcation.t(), load_opts()) :: :ok
+  def load_transaction(tx = %Transaction{}, opts \\ []) do
+    io_transaction? = Keyword.get(opts, :io_transaction?, false)
+    authorized_nodes = P2P.authorized_and_available_nodes()
+
+    # Ingest all the movements to fill up the UTXO list
+    ingest_utxo(tx, authorized_nodes)
+
+    consume_utxo(tx, io_transaction?, authorized_nodes)
+
+    Logger.info("Loaded into in memory UTXO tables",
+      transaction_address: Base.encode16(tx.address),
+      transaction_type: tx.type
+    )
+  end
+
+  defp ingest_utxo(
+         %Transaction{
+           address: address,
+           validation_stamp: %ValidationStamp{
+             protocol_version: protocol_version,
+             timestamp: timestamp,
+             ledger_operations: %LedgerOperations{transaction_movements: transaction_movements}
+           }
+         },
+         authorized_nodes
+       ) do
+    Enum.each(
+      transaction_movements,
+      fn %TransactionMovement{to: to, amount: amount, type: type} ->
+        genesis_address =
+          case DB.find_genesis_address(to) do
+            {:ok, address} ->
+              address
+
+            _ ->
+              # Support when the resolved address is the genesis address
+              to
+          end
+
+        utxo = %VersionedUnspentOutput{
+          unspent_output: %UnspentOutput{
+            from: address,
+            amount: amount,
+            timestamp: timestamp,
+            type: type
+          },
+          protocol_version: protocol_version
+        }
+
+        if genesis_node?(genesis_address, authorized_nodes) do
+          Loader.add_utxo(utxo, genesis_address)
+        end
+      end
+    )
+  end
+
+  defp consume_utxo(
+         tx = %Transaction{
+           validation_stamp: %ValidationStamp{
+             ledger_operations: %LedgerOperations{consumed_inputs: consumed_inputs}
+           }
+         },
+         io_transaction?,
+         authorized_nodes
+       ) do
+    case find_genesis_address(tx) do
+      {:ok, genesis_address} ->
+        # We need to determine whether the node is responsible of the chain genesis pool as the transaction have been received as an I/O transaction.
+        chain_transaction? =
+          (not io_transaction? or TransactionChain.get_size(genesis_address) > 0) and
```

Comment:
This is because a genesis node might receive the transaction as I/O storage node or as chain storage node (due to the possible election overlap).
Hence if it receives the transaction as I/O (for example as genesis) it can only consume the transaction if it already received some inputs before (meaning stored as I/O transaction).
However if it's an overlap (genesis node == transaction nodes), it can consume the transaction according to the genesis_node? election